content/ui-ux.md

```diff
@@ -0,0 +1,514 @@
+
+To throttle writes, a typical approach is to use underscore's [`.throttle()`](http://underscorejs.org/#throttle) or [`.debounce()`](http://underscorejs.org/#debounce) functions. For instance, in the Todos example app, we throttle writes on user input to 300ms:
+
+```js
+Template.todosItem.helpers({
+  // update the text of the item on keypress but throttle the event to ensure
+  // we don't flood the server with updates (handles the event at most once
+  // every 300ms)
+  'keyup input[type=text]': _.throttle(function(event) {
+    Todos.methods.updateText.call({
+      todoId: this.todo._id,
+      newText: event.target.value
+    }, (err) => {
+      err && alert(err.error);
+    });
+  }, 300)
+});
+```
+
+#### Limiting re-rendering
+
+Even if you aren't saving data over the wire to the database on every user input, sometimes you still may wish to update in-memory data stores on every user change. In theory this is fine, but in practice sometimes if updating that data store triggers a lot of UI changes, you can see poor performance (and missed keystrokes) when you do that. In such cases you can limit re-rendering by throttling such changes in a similar way to the way we throttle the method call above.
+
+#### Scroll events
+
+Another type of event that fires very frequently is a scroll event. Such event trigger as a user scrolls up and down a page and are commonly used for "waypoint" behaviour. For example in the Localmarket example app, we use the [jquery-waypoints](https://atmospherejs.com/meteor/jquery-waypoints) library to trigger a class being attached to the nav element when you scroll past a certain point:
+
+```js
+Template.nav.onRendered(function() {
+  var $nav = this.$('nav');
+  $nav.siblings('.content-scrollable:not(.static-nav)').children().first().waypoint(function(direction) {
+    $nav.toggleClass('scrolled', direction === 'down');
+  }, {
+    context: '.content-scrollable',
+    offset: -200
+  });
+});
+```
+
+An important thing to be careful of is that if you are not using a library such as `jquery-waypooints` to do this, then you need to be careful about not causing too much work to happen in each `scroll` event. If you do so, then users will see poor scrolling performance and your app will not feel performant. You can use a throttle or [`requestAnimationFrame`](https://developer.mozilla.org/en-US/docs/Web/API/window/requestAnimationFrame) to make things smooth if you are careful.
+
+
+## User Experience Patterns
+
+There are some common user experience (UX---as in how your app behaves) patterns that are typical to most Meteor apps that are worth exploring here. These patterns relate heavily to the way the data they are allow the user is interacting with is subscribed to and published, so there are similar sections in the {data loading article} which deal with the data side of them.
```

Comment:
I think there's some kind of typo in the sentence: "These patterns relate heavily to the way the data they are allow the user is interacting with is subscribed to and published"